content/posts/vault-migration/index.md

```diff
@@ -0,0 +1,558 @@
+   capabilities = [ "create", "read", "update", "list" ]
+  }
+  # Policy allowing to restore vault's snapshots
+  path "sys/storage/raft/snapshot-force" {
+   capabilities = [ "create", "read", "update", "list" ]
+  }
+  # Policy allowing to seal a vault
+  path "sys/seal" {
+   capabilities = ["update", "sudo"]
+  }
+  ```
+
+- Detect the index of the leader pod in each Vault instance
+
+  ```bash
+  kubectl get pod -l "vault-active=true" -o jsonpath \
+    --template "{.items[0].metadata.labels.apps\.kubernetes\.io/pod-index}"
+  ```
+
+- Create tokens for each vault instances
+
+  ```bash
+  switch_context "${cluster_name}" "${vault_name}" > /dev/null
+  leader_index="$(detect_vault_leader_index)" > /dev/null
+
+  kubectl cp migration-policy.hcl "${vault_name}-${leader_index}":/tmp > /dev/null
+
+  kubectl exec -it "${vault_name}-${leader_index}" -- \
+    vault login "${admin_token}" > /dev/null
+
+  kubectl exec -it "${vault_name}-${leader_index}" -- \
+    sh -c "cat /tmp/migration-policy.hcl | vault policy write migration -" \
+    > /dev/null
+
+  kubectl exec -it "${vault_name}-${leader_index}" -- \
+    vault token create -policy=migration -period=30m -format json \
+    | jq -r '.auth.client_token'
+
+  kubectl exec -it "${vault_name}-${leader_index}" -- \
+    rm /tmp/migration-policy.hcl > /dev/null
+  ```
+
+### 3. Block access to the old Vault instance
+
+Before creating the snapshot, it’s important to prevent any writes to the old
+Vault instance. This step ensures no data is changed after the snapshot and
+avoids issues with lease revocation.
+
+- Backup and delete the ingress for the old Vault instance
+
+  ```bash
+  kubectl get ing vault -o yaml > "${BACKUPS_DIR}/${OLD_VAULT_INGRESS_FILENAME}"
+
+  kubectl delete ing vault --wait=true
+  ```
+
+### 4. Create snapshot of the old Vault instance
+
+We now take a snapshot of the Vault data using the Raft backend's built-in
+snapshot mechanism. After the snapshot is saved and downloaded, we immediately
+disable the old Vault to avoid unintended behavior.
+
+- Log in with the migration token
+
+  ```bash
+  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
+    vault login "${OLD_VAULT_ADMIN_TOKEN}"
+  ```
+
+- Create the snapshot in the leader pod
+
+  ```bash
+  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
+    vault operator raft snapshot save "${snapshot_filepath}"
+  ```
+
+- Dowload the snapshot
+
+  ```bash
+  kubectl cp "${OLD_VAULT_NAME}-${leader_index}":"${snapshot_filepath}" \
+    "${BACKUPS_DIR}/${OLD_VAULT_SNAPSHOT_FILENAME}"
+
+  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
+    rm "${snapshot_temp_filepath}"
+  ```
+
+- Disable the old Vault instance by changing the command of the Vault
+  StatefulSet to `sleep`
+
+  ```bash
+  kubectl get statefulset vault -o yaml > \
+  "${BACKUPS_DIR}/${OLD_VAULT_STATEFULSET_FILENAME}"
+  cp "${BACKUPS_DIR}/${OLD_VAULT_STATEFULSET_FILENAME}" .
+
+  ed "${OLD_VAULT_STATEFULSET_FILENAME}" <<EOF
```

Comment:
Here we should show before and after states of the modified file.

Reply:
@theseion it's probably easier for you to grab the yaml file from the actual environment. I no longer have access. Can you take care of this?

Reply:
I will do the same for the Vault ConfigMap in section 6.